README.md
Forked for debugging do not use

# Azure AMQP Common
[![Go Report Card](https://goreportcard.com/badge/github.com/jackjameshoward/azure-amqp-common-go)](https://goreportcard.com/report/github.com/jackjameshoward/azure-amqp-common-go)
[![godoc](https://godoc.org/github.com/jackjameshoward/azure-amqp-common-go?status.svg)](https://godoc.org/github.com/jackjameshoward/azure-amqp-common-go)
[![Build Status](https://travis-ci.org/Azure/azure-amqp-common-go.svg?branch=master)](https://travis-ci.org/Azure/azure-amqp-common-go)

This project contains reusable components for AMQP based services like Event Hub and Service Bus. You will find 
abstractions over authentication, claims-based security, connection string parsing and RPC for AMQP.

If you are looking for the Azure Event Hub library for go, you can find it [here](https://aka.ms/azure-event-hubs-go).

If you are looking for the Azure Service Bus library for go, you can find it [here](https://aka.ms/azure-service-bus-go).

## Install with Go modules
If you want to use stable versions of the library, please use Go modules.

**NOTE**: versions prior to 3.0.0 depend on pack.ag/amqp which is no longer maintained.
Any new code should not use versions prior to 3.0.0.

### Using go get targeting version 4.x.x
``` bash
go get github.com/jackjameshoward/azure-amqp-common-go
```

### Using go get targeting version 3.x.x
``` bash
go get github.com/jackjameshoward/azure-amqp-common-go/v3
```

## Contributing

This project welcomes contributions and suggestions.  Most contributions require you to agree to a
Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us
the rights to use your contribution. For details, visit https://cla.microsoft.com.

When you submit a pull request, a CLA-bot will automatically determine whether you need to provide
a CLA and decorate the PR appropriately (e.g., label, comment). Simply follow the instructions
provided by the bot. You will only need to do this once across all repos using our CLA.

This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or
contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

## License

MIT, see [LICENSE](./LICENSE).

## Contribute

See [CONTRIBUTING.md](.github/CONTRIBUTING.md).


changelog.md
# Change Log

## `v4.2.0`
- Update to the GA verison of go-amqp

## `v4.1.0`
- Update to the latest go-amqp
  [PR#72](https://github.com/jackjameshoward/azure-amqp-common-go/pull/72)

## `v4.0.0`
- Updated to the latest go-amqp which includes a few minor changes in public surface area.
  [PR#68](https://github.com/jackjameshoward/azure-amqp-common-go/pull/68)

## `v3.2.0`
- Change the default credits for the RPC link to be more reasonable (1000)
  [PR#54](https://github.com/jackjameshoward/azure-amqp-common-go/pull/54)

## `v3.1.2`
- Fixing a potential race condition when an RPC link is shut down while still sending requests
  or handling responses.
  [PR#55](https://github.com/jackjameshoward/azure-amqp-common-go/pull/55)
- Upgrading to go-amqp v0.13.13, which fixes an issue with simultaneous settling on the rpc link.

## `v3.1.1`
- Change `Link` so it can handle parallel requests. 
  [PR#52](https://github.com/jackjameshoward/azure-amqp-common-go/pull/52)

## `v3.1.0`
- Add support for passing managed identity user-assigned client ID.

## `v3.0.1`
- add context to message deposition methods
- update dependencies

## `v3.0.0`
- switch from pack.ag/amqp to github.com/Azure/go-amqp
- bump major version

## `v2.1.1`
- bump amqp to v0.12.1
- bump azure sdk for go to v32.5.0
- bump go-autorest

## `v2.1.0`
- add session filters for RPC links
- bump amqp to v0.11.2
- add more logging in RPC operations

## `v2.0.0`
- [**breaking change** remove persist and move into the Event Hubs package](https://github.com/Azure/azure-event-hubs-go/pull/112)
- **breaking change** remove log package in favor of https://github.com/devigned/tab

## `v1.1.4`
- allow status description on RPC calls to be empty without returning an error https://github.com/Azure/azure-event-hubs-go/issues/88

## `v1.1.3`
- adding automatic server-timeout field for `rpc` package. It gleans the appropriate value from the context passed to it

## `v1.1.2`
- adopting go modules 

## `v1.1.1`
- broadening accepted versions of pack.ag/amqp

## `v1.1.0`

- adding the ability to reuse an AMQP session while making RPCs
- bug fixes

## `v1.0.3`
- updating dependencies, adding new 'go-autorest' constraint

## `v1.0.2`
- adding resiliency against malformed "status-code" and "status-description" properties in rpc responses

## `v1.0.1`
- bump version constant

## `v1.0.0`
- moved to opencensus from opentracing
- committing to backward compatibility

## `v0.7.0`
- update AMQP dependency to 0.7.0

## `v0.6.0`
- **Breaking Change** change the parse connection signature and make it more strict
- fix errors imports

## `v0.5.0`
- **Breaking Change** lock dependency to AMQP

## `v0.4.0`
- **Breaking Change** remove namespace from SAS provider and return struct rather than interface 

## `v0.3.2`
- Return error on retry. Was returning nil if not retryable.

## `v0.3.1`
- Fix missing defer on spans

## `v0.3.0`
- add opentracing support
- upgrade amqp to pull in the changes where close accepts context (breaking change)

## `v0.2.4`
- connection string keys are case insensitive 

## `v0.2.3`
- handle remove trailing slash from host

## `v0.2.2`
- handle connection string values which contain `=`

## `v0.2.1`
- parse connection strings using key / values rather than regex

## `v0.2.0`
- add file checkpoint persister

## `v0.1.0`
- initial release

rpc/rpc_test.go
package rpc

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"testing"
	"time"

	"github.com/Azure/go-amqp"
	"github.com/jackjameshoward/azure-amqp-common-go/uuid"
	"github.com/stretchr/testify/require"
)

func TestResponseRouterBasic(t *testing.T) {
	receiver := &fakeReceiver{
		Responses: []rpcResponse{
			{amqpMessageWithCorrelationId("my message id"), nil},
			{nil, &amqp.LinkError{}},
		},
	}

	link := &Link{
		responseMap: map[string]chan rpcResponse{
			"my message id": make(chan rpcResponse, 1),
		},
		receiver: receiver,
	}

	ch := link.responseMap["my message id"]

	link.startResponseRouter()
	result := <-ch
	require.EqualValues(t, result.message.Data[0], []byte("ID was my message id"))
	require.Empty(t, receiver.Responses)
	require.Nil(t, link.responseMap, "Response map is nil'd out after we get a closed error")
}

func TestResponseRouterMissingMessageID(t *testing.T) {
	receiver := &fakeReceiver{
		Responses: []rpcResponse{
			{amqpMessageWithCorrelationId("my message id"), nil},
			{nil, &amqp.LinkError{}},
		},
	}

	link := &Link{
		responseMap: map[string]chan rpcResponse{},
		receiver:    receiver,
	}

	link.startResponseRouter()
	require.Empty(t, receiver.Responses)
}

func TestResponseRouterBadCorrelationID(t *testing.T) {
	messageWithBadCorrelationID := &amqp.Message{
		Properties: &amqp.MessageProperties{
			CorrelationID: uint64(1),
		},
	}

	receiver := &fakeReceiver{
		Responses: []rpcResponse{
			{messageWithBadCorrelationID, nil},
			{nil, &amqp.LinkError{}},
		},
	}

	link := &Link{
		responseMap: map[string]chan rpcResponse{},
		receiver:    receiver,
	}

	link.startResponseRouter()
	require.Empty(t, receiver.Responses)
}

func TestResponseRouterFatalErrors(t *testing.T) {
	fatalErrors := []error{
		&amqp.LinkError{},
		&amqp.ConnError{},
		&amqp.SessionError{},
	}

	for _, fatalError := range fatalErrors {
		t.Run(fatalError.Error(), func(t *testing.T) {
			receiver := &fakeReceiver{
				Responses: []rpcResponse{
					{nil, fatalError},
				},
			}
			sentinelCh := make(chan rpcResponse, 1)

			link := &Link{
				responseMap: map[string]chan rpcResponse{
					"sentinel": sentinelCh,
				},
				receiver: receiver,
			}

			link.startResponseRouter()
			require.Empty(t, receiver.Responses)

			// also, we should have broadcasted that the link is closed to anyone else
			// that had not yet received a response but was still waiting.
			select {
			case rpcResponse := <-sentinelCh:
				require.Error(t, rpcResponse.err, fatalError.Error())
				require.Nil(t, rpcResponse.message)
			case <-time.After(time.Second * 5):
				require.Fail(t, "sentinel channel should have received a message")
			}
		})
	}
}

func TestResponseRouterNoResponse(t *testing.T) {
	receiver := &fakeReceiver{
		Responses: []rpcResponse{
			{nil, errors.New("Some other error that will get ignored since we can't route it to anyone (ie: no message ID)")},
			{nil, &amqp.ConnError{}},
		},
	}

	link := &Link{
		responseMap: map[string]chan rpcResponse{},
		receiver:    receiver,
	}

	link.startResponseRouter()
	require.Empty(t, receiver.Responses)
}

func TestAddMessageID(t *testing.T) {
	message, id, err := addMessageID(&amqp.Message{}, uuid.NewV4)
	require.NoError(t, err)
	require.NotEmpty(t, id)
	require.EqualValues(t, message.Properties.MessageID, id)

	m := &amqp.Message{
		Data: [][]byte{[]byte("hello world")},
		Properties: &amqp.MessageProperties{
			UserID:    []byte("my user ID"),
			MessageID: "is that will not be copied"},
	}
	message, id, err = addMessageID(m, uuid.NewV4)
	require.NoError(t, err)
	require.NotEmpty(t, id)
	require.EqualValues(t, message.Properties.MessageID, id)
	require.EqualValues(t, message.Properties.UserID, []byte("my user ID"))
	require.EqualValues(t, message.Data[0], []byte("hello world"))
}

func TestRPCBasic(t *testing.T) {
	fakeUUID := uuid.UUID([16]byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15})
	replyMessage := amqpMessageWithCorrelationId(fakeUUID.String())
	replyMessage.ApplicationProperties = map[string]interface{}{
		"status-code": int32(200),
	}

	ch := make(chan struct{})
	sender := &fakeSender{ch: ch}
	receiver := &fakeReceiver{
		Responses: []rpcResponse{
			{replyMessage, nil},
			{nil, &amqp.ConnError{}},
		},
		ch: ch,
	}

	l := &Link{
		receiver:                receiver,
		sender:                  sender,
		startResponseRouterOnce: &sync.Once{},
		responseMap:             map[string]chan rpcResponse{},

		uuidNewV4: func() (uuid.UUID, error) {
			return fakeUUID, nil
		},
		messageAccept: func(ctx context.Context, message *amqp.Message) error {
			return nil
		},
	}

	messageToSend := &amqp.Message{}
	resp, err := l.RPC(context.Background(), messageToSend)
	require.NoError(t, err)

	require.EqualValues(t, fakeUUID.String(), sender.Sent[0].Properties.MessageID, "Sent message contains a uniquely generated ID")
	require.EqualValues(t, fakeUUID.String(), resp.Message.Properties.CorrelationID, "Correlation ID matches our originally sent message")

	require.Nil(t, replyMessage.Properties.MessageID, "Original message not modified")
}

func TestRPCFailedSend(t *testing.T) {
	// important bit is that we clean up the channel we stored in the map
	// since we're no longer waiting for the response.
	fakeUUID := uuid.UUID([16]byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15})
	replyMessage := amqpMessageWithCorrelationId(fakeUUID.String())
	replyMessage.ApplicationProperties = map[string]interface{}{
		"status-code": int32(200),
	}

	ch := make(chan struct{})

	sender := &fakeSender{
		ch: ch,
	}
	receiver := &fakeReceiver{
		ch: ch,
		Responses: []rpcResponse{
			{nil, &amqp.ConnError{}},
		},
	}

	l := &Link{
		receiver:                receiver,
		sender:                  sender,
		startResponseRouterOnce: &sync.Once{},
		responseMap:             map[string]chan rpcResponse{},

		uuidNewV4: func() (uuid.UUID, error) {
			return fakeUUID, nil
		},
		messageAccept: func(ctx context.Context, message *amqp.Message) error {
			panic("Should not be called")
		},
	}

	messageToSend := &amqp.Message{}
	cancelledContext, cancel := context.WithCancel(context.Background())
	cancel()

	resp, err := l.RPC(cancelledContext, messageToSend)
	require.Nil(t, resp)
	require.EqualError(t, err, context.Canceled.Error())

	require.EqualValues(t, fakeUUID.String(), sender.Sent[0].Properties.MessageID, "Sent message contains a uniquely generated ID")
}

func TestRPCNilMessageMap(t *testing.T) {
	fakeSender := &fakeSender{}
	fakeReceiver := &fakeReceiver{
		Responses: []rpcResponse{
			// this should let us see what deleteChannelFromMap does
			{amqpMessageWithCorrelationId("hello"), nil},
			{nil, &amqp.LinkError{}},
		},
	}

	link := &Link{
		sender:   fakeSender,
		receiver: fakeReceiver,
		// responseMap is nil if the broadcastError() function is called. Since this can be
		// at any time our individual map functions need to handle the map not being
		// there.
		responseMap:             nil,
		startResponseRouterOnce: &sync.Once{},
		uuidNewV4:               uuid.NewV4,
	}

	// sanity check - all the map/channel functions are returning nil
	require.Nil(t, link.addChannelToMap("hello"))
	require.Nil(t, link.deleteChannelFromMap("hello"))

	link.startResponseRouter()

	require.Empty(t, fakeReceiver.Responses, "All responses are used")

	// we're not testing the responseRouter for this second part, so just short-circuit
	// the running.
	link.startResponseRouterOnce.Do(func() {})

	// now check that sending can handle it.
	resp, err := link.RPC(context.Background(), &amqp.Message{})
	var linkErr *amqp.LinkError
	require.ErrorAs(t, err, &linkErr)
	require.Nil(t, resp)
}

func amqpMessageWithCorrelationId(id string) *amqp.Message {
	return &amqp.Message{
		Data: [][]byte{[]byte(fmt.Sprintf("ID was %s", id))},
		Properties: &amqp.MessageProperties{
			CorrelationID: id,
		},
	}
}

type fakeReceiver struct {
	Responses []rpcResponse
	ch        <-chan struct{}
}

func (fr *fakeReceiver) Receive(ctx context.Context, o *amqp.ReceiveOptions) (*amqp.Message, error) {
	// wait until the actual send if we're simulating request/response
	if fr.ch != nil {
		<-fr.ch
	}

	resp := fr.Responses[0]
	fr.Responses = fr.Responses[1:]
	return resp.message, resp.err
}

func (fr *fakeReceiver) Close(ctx context.Context) error {
	panic("Not used for this test")
}

type fakeSender struct {
	Sent []*amqp.Message
	ch   chan<- struct{}
}

func (s *fakeSender) Send(ctx context.Context, msg *amqp.Message, o *amqp.SendOptions) error {
	s.Sent = append(s.Sent, msg)

	if s.ch != nil {
		close(s.ch)
	}

	return nil
}

func (fs *fakeSender) Close(ctx context.Context) error {
	panic("Not used for this test")
}


cbs/cbs.go
// Package cbs provides the functionality for negotiating claims-based security over AMQP for use in Azure Service Bus
// and Event Hubs.
package cbs

//	MIT License
//
//	Copyright (c) Microsoft Corporation. All rights reserved.
//
//	Permission is hereby granted, free of charge, to any person obtaining a copy
//	of this software and associated documentation files (the "Software"), to deal
//	in the Software without restriction, including without limitation the rights
//	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//	copies of the Software, and to permit persons to whom the Software is
//	furnished to do so, subject to the following conditions:
//
//	The above copyright notice and this permission notice shall be included in all
//	copies or substantial portions of the Software.
//
//	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//	SOFTWARE

import (
	"context"
	"fmt"
	"time"

	"github.com/devigned/tab"

	"github.com/Azure/go-amqp"
	"github.com/jackjameshoward/azure-amqp-common-go/auth"
	"github.com/jackjameshoward/azure-amqp-common-go/internal/tracing"
	"github.com/jackjameshoward/azure-amqp-common-go/rpc"
)

const (
	cbsAddress           = "$cbs"
	cbsOperationKey      = "operation"
	cbsOperationPutToken = "put-token"
	cbsTokenTypeKey      = "type"
	cbsAudienceKey       = "name"
	cbsExpirationKey     = "expiration"
)

// NegotiateClaim attempts to put a token to the $cbs management endpoint to negotiate auth for the given audience
func NegotiateClaim(ctx context.Context, audience string, conn *amqp.Conn, provider auth.TokenProvider) error {
	ctx, span := tracing.StartSpanFromContext(ctx, "az-amqp-common.cbs.NegotiateClaim")
	defer span.End()

	link, err := rpc.NewLink(ctx, conn, cbsAddress)
	if err != nil {
		tab.For(ctx).Error(err)
		return err
	}
	defer func() {
		if err := link.Close(ctx); err != nil {
			tab.For(ctx).Error(err)
		}
	}()

	token, err := provider.GetToken(audience)
	if err != nil {
		tab.For(ctx).Error(err)
		return err
	}

	tab.For(ctx).Debug(fmt.Sprintf("negotiating claim for audience %s with token type %s and expiry of %s", audience, token.TokenType, token.Expiry))
	msg := &amqp.Message{
		Value: token.Token,
		ApplicationProperties: map[string]interface{}{
			cbsOperationKey:  cbsOperationPutToken,
			cbsTokenTypeKey:  string(token.TokenType),
			cbsAudienceKey:   audience,
			cbsExpirationKey: token.Expiry,
		},
	}

	res, err := link.RetryableRPC(ctx, 3, 1*time.Second, msg)
	if err != nil {
		tab.For(ctx).Error(err)
		return err
	}

	tab.For(ctx).Debug(fmt.Sprintf("negotiated with response code %d and message: %s", res.Code, res.Description))
	return nil
}


internal/tracing/tracing.go
package tracing

import (
	"context"
	"os"

	"github.com/devigned/tab"
	common "github.com/jackjameshoward/azure-amqp-common-go/internal"
)

// StartSpanFromContext starts a span given a context and applies common library information
func StartSpanFromContext(ctx context.Context, operationName string) (context.Context, tab.Spanner) {
	ctx, span := tab.StartSpan(ctx, operationName)
	ApplyComponentInfo(span)
	return ctx, span
}

// ApplyComponentInfo applies eventhub library and network info to the span
func ApplyComponentInfo(span tab.Spanner) {
	span.AddAttributes(
		tab.StringAttribute("component", "github.com/jackjameshoward/azure-amqp-common-go"),
		tab.StringAttribute("version", common.Version))
	applyNetworkInfo(span)
}

func applyNetworkInfo(span tab.Spanner) {
	hostname, err := os.Hostname()
	if err == nil {
		span.AddAttributes(tab.StringAttribute("peer.hostname", hostname))
	}
}


rpc/rpc.go
// Package rpc provides functionality for request / reply messaging. It is used by package mgmt and cbs.
package rpc

//	MIT License
//
//	Copyright (c) Microsoft Corporation. All rights reserved.
//
//	Permission is hereby granted, free of charge, to any person obtaining a copy
//	of this software and associated documentation files (the "Software"), to deal
//	in the Software without restriction, including without limitation the rights
//	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//	copies of the Software, and to permit persons to whom the Software is
//	furnished to do so, subject to the following conditions:
//
//	The above copyright notice and this permission notice shall be included in all
//	copies or substantial portions of the Software.
//
//	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//	SOFTWARE

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/devigned/tab"

	"github.com/Azure/go-amqp"
	common "github.com/jackjameshoward/azure-amqp-common-go"
	"github.com/jackjameshoward/azure-amqp-common-go/internal/tracing"
	"github.com/jackjameshoward/azure-amqp-common-go/uuid"
)

const (
	replyPostfix           = "-reply-to-"
	statusCodeKey          = "status-code"
	descriptionKey         = "status-description"
	defaultReceiverCredits = 1000
)

type (
	// Link is the bidirectional communication structure used for CBS negotiation
	Link struct {
		session *amqp.Session

		receiver amqpReceiver // *amqp.Receiver
		sender   amqpSender   // *amqp.Sender

		clientAddress string
		sessionID     *string
		useSessionID  bool
		id            string

		responseMu              sync.Mutex
		startResponseRouterOnce *sync.Once
		responseMap             map[string]chan rpcResponse

		// for unit tests
		uuidNewV4     func() (uuid.UUID, error)
		messageAccept func(ctx context.Context, message *amqp.Message) error
	}

	// Response is the simplified response structure from an RPC like call
	Response struct {
		Code        int
		Description string
		Message     *amqp.Message
	}

	// LinkOption provides a way to customize the construction of a Link
	LinkOption func(link *Link) error

	rpcResponse struct {
		message *amqp.Message
		err     error
	}

	// Actually: *amqp.Receiver
	amqpReceiver interface {
		Receive(ctx context.Context, o *amqp.ReceiveOptions) (*amqp.Message, error)
		Close(ctx context.Context) error
	}

	amqpSender interface {
		Send(ctx context.Context, msg *amqp.Message, o *amqp.SendOptions) error
		Close(ctx context.Context) error
	}
)

// LinkWithSessionFilter configures a Link to use a session filter
func LinkWithSessionFilter(sessionID *string) LinkOption {
	return func(l *Link) error {
		l.sessionID = sessionID
		l.useSessionID = true
		return nil
	}
}

// NewLink will build a new request response link
func NewLink(ctx context.Context, conn *amqp.Conn, address string, opts ...LinkOption) (*Link, error) {
	authSession, err := conn.NewSession(ctx, nil)
	if err != nil {
		return nil, err
	}

	return NewLinkWithSession(ctx, authSession, address, opts...)
}

// NewLinkWithSession will build a new request response link, but will reuse an existing AMQP session
func NewLinkWithSession(ctx context.Context, session *amqp.Session, address string, opts ...LinkOption) (*Link, error) {
	linkID, err := uuid.NewV4()
	if err != nil {
		return nil, err
	}

	id := linkID.String()
	link := &Link{
		session:       session,
		clientAddress: strings.Replace("$", "", address, -1) + replyPostfix + id,
		id:            id,

		uuidNewV4:               uuid.NewV4,
		responseMap:             map[string]chan rpcResponse{},
		startResponseRouterOnce: &sync.Once{},
	}

	for _, opt := range opts {
		if err := opt(link); err != nil {
			return nil, err
		}
	}

	sender, err := session.NewSender(ctx, address, nil)
	if err != nil {
		return nil, err
	}

	receiverOpts := amqp.ReceiverOptions{
		Credit:        defaultReceiverCredits,
		TargetAddress: link.clientAddress,
	}

	if link.sessionID != nil {
		const name = "com.microsoft:session-filter"
		const code = uint64(0x00000137000000C)
		if link.sessionID == nil {
			receiverOpts.Filters = append(receiverOpts.Filters, amqp.NewLinkFilter(name, code, nil))
		} else {
			receiverOpts.Filters = append(receiverOpts.Filters, amqp.NewLinkFilter(name, code, link.sessionID))
		}
	}

	receiver, err := session.NewReceiver(ctx, address, &receiverOpts)
	if err != nil {
		// make sure we close the sender
		clsCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
		defer cancel()

		_ = sender.Close(clsCtx)
		return nil, err
	}

	link.sender = sender
	link.receiver = receiver
	link.messageAccept = receiver.AcceptMessage

	return link, nil
}

// RetryableRPC attempts to retry a request a number of times with delay
func (l *Link) RetryableRPC(ctx context.Context, times int, delay time.Duration, msg *amqp.Message) (*Response, error) {
	ctx, span := tracing.StartSpanFromContext(ctx, "az-amqp-common.rpc.RetryableRPC")
	defer span.End()

	res, err := common.Retry(times, delay, func() (interface{}, error) {
		ctx, span := tracing.StartSpanFromContext(ctx, "az-amqp-common.rpc.RetryableRPC.retry")
		defer span.End()

		res, err := l.RPC(ctx, msg)

		if err != nil {
			tab.For(ctx).Error(fmt.Errorf("error in RPC via link %s: %v", l.id, err))
			return nil, err
		}

		switch {
		case res.Code >= 200 && res.Code < 300:
			tab.For(ctx).Debug(fmt.Sprintf("successful rpc on link %s: status code %d and description: %s", l.id, res.Code, res.Description))
			return res, nil
		case res.Code >= 500:
			errMessage := fmt.Sprintf("server error link %s: status code %d and description: %s", l.id, res.Code, res.Description)
			tab.For(ctx).Error(errors.New(errMessage))
			return nil, common.Retryable(errMessage)
		default:
			errMessage := fmt.Sprintf("unhandled error link %s: status code %d and description: %s", l.id, res.Code, res.Description)
			tab.For(ctx).Error(errors.New(errMessage))
			return nil, common.Retryable(errMessage)
		}
	})
	if err != nil {
		tab.For(ctx).Error(err)
		return nil, err
	}
	return res.(*Response), nil
}

// startResponseRouter is responsible for taking any messages received on the 'response'
// link and forwarding it to the proper channel. The channel is being select'd by the
// original `RPC` call.
func (l *Link) startResponseRouter() {
	for {
		res, err := l.receiver.Receive(context.Background(), nil)

		// You'll see this when the link is shutting down (either
		// service-initiated via 'detach' or a user-initiated shutdown)
		if isClosedError(err) {
			l.broadcastError(err)
			break
		} else if err != nil {
			// this is some transient error, sleep before trying again
			time.Sleep(time.Second)
		}

		// I don't believe this should happen. The JS version of this same code
		// ignores errors as well since responses should always be correlated
		// to actual send requests. So this is just here for completeness.
		if res == nil {
			continue
		}

		autogenMessageId, ok := res.Properties.CorrelationID.(string)

		if !ok {
			// TODO: it'd be good to track these in some way. We don't have a good way to
			// forward this on at this point.
			continue
		}

		ch := l.deleteChannelFromMap(autogenMessageId)

		if ch != nil {
			ch <- rpcResponse{message: res, err: err}
		}
	}
}

// RPC sends a request and waits on a response for that request
func (l *Link) RPC(ctx context.Context, msg *amqp.Message) (*Response, error) {
	l.startResponseRouterOnce.Do(func() {
		go l.startResponseRouter()
	})

	copiedMessage, messageID, err := addMessageID(msg, l.uuidNewV4)

	if err != nil {
		return nil, err
	}

	// use the copiedMessage from this point
	msg = copiedMessage

	const altStatusCodeKey, altDescriptionKey = "statusCode", "statusDescription"

	ctx, span := tracing.StartSpanFromContext(ctx, "az-amqp-common.rpc.RPC")
	defer span.End()

	msg.Properties.ReplyTo = &l.clientAddress

	if msg.ApplicationProperties == nil {
		msg.ApplicationProperties = make(map[string]interface{})
	}

	if _, ok := msg.ApplicationProperties["server-timeout"]; !ok {
		if deadline, ok := ctx.Deadline(); ok {
			msg.ApplicationProperties["server-timeout"] = uint(time.Until(deadline) / time.Millisecond)
		}
	}

	responseCh := l.addChannelToMap(messageID)

	if responseCh == nil {
		return nil, &amqp.LinkError{}
	}

	err = l.sender.Send(ctx, msg, nil)

	if err != nil {
		l.deleteChannelFromMap(messageID)
		tab.For(ctx).Error(err)
		return nil, err
	}

	var res *amqp.Message

	select {
	case <-ctx.Done():
		l.deleteChannelFromMap(messageID)
		res, err = nil, ctx.Err()
	case resp := <-responseCh:
		// this will get triggered by the loop in 'startReceiverRouter' when it receives
		// a message with our autoGenMessageID set in the correlation_id property.
		res, err = resp.message, resp.err
	}

	if err != nil {
		tab.For(ctx).Error(err)
		return nil, err
	}

	var statusCode int
	statusCodeCandidates := []string{statusCodeKey, altStatusCodeKey}
	for i := range statusCodeCandidates {
		if rawStatusCode, ok := res.ApplicationProperties[statusCodeCandidates[i]]; ok {
			if cast, ok := rawStatusCode.(int32); ok {
				statusCode = int(cast)
				break
			} else {
				err := errors.New("status code was not of expected type int32")
				tab.For(ctx).Error(err)
				return nil, err
			}
		}
	}
	if statusCode == 0 {
		err := errors.New("status codes was not found on rpc message")
		tab.For(ctx).Error(err)
		return nil, err
	}

	var description string
	descriptionCandidates := []string{descriptionKey, altDescriptionKey}
	for i := range descriptionCandidates {
		if rawDescription, ok := res.ApplicationProperties[descriptionCandidates[i]]; ok {
			if description, ok = rawDescription.(string); ok || rawDescription == nil {
				break
			} else {
				return nil, errors.New("status description was not of expected type string")
			}
		}
	}

	span.AddAttributes(tab.StringAttribute("http.status_code", fmt.Sprintf("%d", statusCode)))

	response := &Response{
		Code:        int(statusCode),
		Description: description,
		Message:     res,
	}

	if err := l.messageAccept(ctx, res); err != nil {
		tab.For(ctx).Error(err)
		return response, err
	}

	return response, err
}

// Close the link receiver, sender and session
func (l *Link) Close(ctx context.Context) error {
	ctx, span := tracing.StartSpanFromContext(ctx, "az-amqp-common.rpc.Close")
	defer span.End()

	if err := l.closeReceiver(ctx); err != nil {
		_ = l.closeSender(ctx)
		_ = l.closeSession(ctx)
		return err
	}

	if err := l.closeSender(ctx); err != nil {
		_ = l.closeSession(ctx)
		return err
	}

	return l.closeSession(ctx)
}

func (l *Link) closeReceiver(ctx context.Context) error {
	ctx, span := tracing.StartSpanFromContext(ctx, "az-amqp-common.rpc.closeReceiver")
	defer span.End()

	if l.receiver != nil {
		return l.receiver.Close(ctx)
	}
	return nil
}

func (l *Link) closeSender(ctx context.Context) error {
	ctx, span := tracing.StartSpanFromContext(ctx, "az-amqp-common.rpc.closeSender")
	defer span.End()

	if l.sender != nil {
		return l.sender.Close(ctx)
	}
	return nil
}

func (l *Link) closeSession(ctx context.Context) error {
	ctx, span := tracing.StartSpanFromContext(ctx, "az-amqp-common.rpc.closeSession")
	defer span.End()

	if l.session != nil {
		return l.session.Close(ctx)
	}
	return nil
}

// addChannelToMap adds a channel which will be used by the response router to
// notify when there is a response to the request.
// If l.responseMap is nil (for instance, via broadcastError) this function will
// return nil.
func (l *Link) addChannelToMap(messageID string) chan rpcResponse {
	l.responseMu.Lock()
	defer l.responseMu.Unlock()

	if l.responseMap == nil {
		return nil
	}

	responseCh := make(chan rpcResponse, 1)
	l.responseMap[messageID] = responseCh

	return responseCh
}

// deleteChannelFromMap removes the message from our internal map and returns
// a channel that the corresponding RPC() call is waiting on.
// If l.responseMap is nil (for instance, via broadcastError) this function will
// return nil.
func (l *Link) deleteChannelFromMap(messageID string) chan rpcResponse {
	l.responseMu.Lock()
	defer l.responseMu.Unlock()

	if l.responseMap == nil {
		return nil
	}

	ch := l.responseMap[messageID]
	delete(l.responseMap, messageID)

	return ch
}

// broadcastError notifies the anyone waiting for a response that the link/session/connection
// has closed.
func (l *Link) broadcastError(err error) {
	l.responseMu.Lock()
	defer l.responseMu.Unlock()

	for _, ch := range l.responseMap {
		ch <- rpcResponse{err: err}
	}

	l.responseMap = nil
}

// addMessageID generates a unique UUID for the message. When the service
// responds it will fill out the correlation ID property of the response
// with this ID, allowing us to link the request and response together.
//
// NOTE: this function copies 'message', adding in a 'Properties' object
// if it does not already exist.
func addMessageID(message *amqp.Message, uuidNewV4 func() (uuid.UUID, error)) (*amqp.Message, string, error) {
	uuid, err := uuidNewV4()

	if err != nil {
		return nil, "", err
	}

	autoGenMessageID := uuid.String()

	// we need to modify the message so we'll make a copy
	copiedMessage := *message

	if message.Properties == nil {
		copiedMessage.Properties = &amqp.MessageProperties{
			MessageID: autoGenMessageID,
		}
	} else {
		// properties already exist, make a copy and then update
		// the message ID
		copiedProperties := *message.Properties
		copiedProperties.MessageID = autoGenMessageID

		copiedMessage.Properties = &copiedProperties
	}

	return &copiedMessage, autoGenMessageID, nil
}

func isClosedError(err error) bool {
	var connError *amqp.ConnError
	var sessionError *amqp.SessionError
	var linkError *amqp.LinkError

	return (errors.As(err, &linkError) && linkError.RemoteErr == nil) ||
		errors.As(err, &sessionError) ||
		errors.As(err, &connError)
}
